4주차 과제 - Todo 리스트 Redux를 사용하여 리팩터링 하기

sejineeee opened the pull request:
안녕하세요!! 3일차인 지금 첫 PR을 남겨봅니다😢
아직 Redux도 어렵기도 하고, test 코드를 작성하는 것도 어려워서 첫 번째 과제는 아샬님의 강의 영상을 보고, 해보았습니다.
commit 메시지를 작성하다 보니 시간이 촉박하여 coverage 확인을 하지 못 했습니다.
아직 모르는 부분이 너무 많아서 어느 부분을 모르는 지 아직은 잘 모르겠어요!!
이번 한 주도 잘 부탁드립니다!!!

sejineeee committed:
chore: react-redux 설치

redux react-redux 설치

sejineeee committed:
feat: store 파일 생성

redux에서 store를 관리할 파일 생성
reducer 분리하여 파일 불러와서 createStore에 인자로 전달

sejineeee committed:
reducer test 파일 생성

reducer test 파일 생성
context 구문 사용하기 위해서 context import
describe-context-it 적용하여 할 일이 있을 경우 없을 경우 나누어 테스트

sejineeee committed:
feat: reducer 파일 추가

초기 상태를 선언하고, reducer에 state의 초기값으로 할당
각 action의 type이 일치할 경우 반환하도로 구현

sejineeee committed:
feat: ListContainer 파일, test 파일 추가

List에 상태를 관리할 수 있도록 ListContainer 파일 생성
ListContainer.test 파일 생성

sejineeee committed:
feat: InputContainer 파일, test 파일 추가

InputContainer 파일 추가
InputContainer.test 파일 추가

sejineeee committed:
test에서 사용할 mock function 파일 추가

test에서 사용할 useDispatch, useSelector mock function 추가

sejineeee committed:
feat: 추가, 변경, 삭제 등 각 action에 대한 함수 파일 생성

할 일을 변경하는 changeTaskTitle action 함수
할 일을 추가하는 addTask action 함수
할 일을 삭제하는 deleteTask acdtion 함수

sejineeee committed:
Page 파일, Page test 파일 삭제

관심사의 분리로 인하여 App 컴포넌트에 ListContainer, InputContainer 호출
그로인하여 Page 파일과 Page test 파일 삭제

sejineeee committed:
refactor: redux를 이용하여 상태관리 분리

redux를 이용하여 상태관리를 분리
상태를 변경해주는 함수는 actions으로 분리하여 App에서는 제거
Page 컴포넌트 호출을 ListContainer와 InputContainer 컴포넌트 호출로 변경

sejineeee committed:
refactor: redux 사용을 위해서 Provider 추가

daadaadaah commented on __mocks__/react-redux.js:
```diff
@@ -0,0 +1,9 @@
+export const useDispatch = jest.fn();
+
+export const useSelector = jest.fn((selector) => selector({
+    tasks: [
+      { id: 1, title: '리덕스 공부하기' },
+      { id: 2, title: 'TDD 공부하기' },
+    ],
+  }),
+);
```
아샬님은 useSelector를 어떻게 mocking 하셨는지 강의를 통해 한번 체크해봅시당!ㅎㅎ

daadaadaah commented on src/reducer.test.js:
```diff
@@ -0,0 +1,84 @@
+          {
+            tasks: [{ id: 1, title: 'Task' }],
+          },
+          deleteTask(100),
```
혹시, 이 테스트는 tasks에 존재하지 않는 id일 경우, 아무런 동작을 하지 않는 걸 기대하고 한 테스트인가요?
만약, 그렇다면, 숫자를 정말 존재할 수 없는 `음수`로 값을 넣어줄 것 같아요! ㅎㅎ진짜 존재하지 않는 숫자라는 것을 드러내기 위해서요!

daadaadaah commented on src/reducer.js:
```diff
@@ -0,0 +1,39 @@
+};
+
+export default function reducer(state = initialState, action) {
+  if (action.type === 'changeTaskTitle') {
+    return {
+      ...state,
+      taskTitle: action.payload.taskTitle,
+    };
+  }
+
+  if (action.type === 'addTask') {
+    const { newId, taskTitle, tasks } = state;
+
+    if (!taskTitle) {
+      return state;
+    }
+
+    return {
+      ...state,
+      newId: newId + 1,
+      taskTitle: '',
+      tasks: [...tasks, { id: newId, title: taskTitle }],
+    };
+  }
+
+  if (action.type === 'deleteTask') {
+    const { tasks } = state;
+    return {
+      ...state,
+      tasks: tasks.filter((task) => task.id !== action.payload.id),
+    };
+  }
```
우리는 이전 과제들을 하면서 if문이 많을 경우, 어떤 걸 활용해서 코드를 간단히 했습니다!ㅎㅎ한번 적용시켜봅시당!ㅎㅎ

daadaadaah commented on src/ListContainer.test.jsx:
```diff
@@ -0,0 +1,24 @@
+
+describe('ListContainer', () => {
+  it('List', () => {
+    const tasks = [
+      { id: 1, title: '리덕스 공부하기' },
+      { id: 2, title: 'TDD 공부하기' },
+    ];
```
우리는 이전 과제들을 하면서 이런 테스트용 데어터를 무엇이라고 했고, 어떻게 관리했었을까요?ㅎㅎ

daadaadaah commented on src/ListContainer.test.jsx:
```diff
@@ -0,0 +1,24 @@
+jest.mock('react-redux');
+
+describe('ListContainer', () => {
+  it('List', () => {
```
describe-(context)-it 에 있는 테스트 내용을 하나로 합쳤을 때, 자연스러운 문장이 나오는게 좋아요!
지금 작성해주신 대로 한문장을 만들어보면, `ListContainer`는 `List`이다. 라는 문장이 만들어집니다.
이렇게 만들어진 문장을 보면, 이 문장이 무엇을 의미하는지 동료가 파악하기 힘들어요! 한문장이 되도록 it문을 수정하면 어떻게 수정하는게 좋을까요?ㅎㅎ(다른 테스트들도 한번 체크해봅시당!ㅎㅎ)

daadaadaah reviewed (changes requested):
이번 PR에서 가장 칭찬해드리고 싶은 부분은 commit 을 잘게 쪼개서 작업단위로 하신 점이예요! 잘하셨어요!👏👏👏
다만, [18fc425](https://github.com/CodeSoom/react-week4-assignment-1/pull/142/commits/18fc425a7d31b2a19e39832e464ca10673efe671) 이 커밋은 redux 사용 전에 먼저 해주셨으면 더 좋았지않을까 생각해요!ㅎㅎ
다음 PR 기다리고 있겠습니당!😊

sejineeee committed:
feat: fixtures 파일 생성

테스트용 데이터를 분리하여 fixtures 폴더에 tasks.js 파일에 추가

sejineeee committed:
review: useSelector mock 함수 수정

강의를 통해서 useSelector mock 함수 수정

sejineeee committed:
review: reducer 함수에 사용했던 if문을 제거하고 객체로 변경

reducers 변수를 만들어, action creator 함수들을 객체에 key: value 형태로 변경

reducer 함수 return문에 action type이 undefined일 경우, state를 반환
삼항연산자 사용하여 구현하였음

sejineeee committed:
review: 입력했을 때, 추가했을 때에 대한 test 추가

입력했을 때, handleChange 작동에 대한 test 추가
렌더링 테스트에 있던 추가 버튼에 대한 것을 추가 버튼 test로 분리

sejineeee committed:
review: 삭제했을 때 테스트 추가

테스트용 데이터를 fixtures로 분리하여 호출해서 사용하는 것으로 변경
dispatch에 대한 mock 함수 추가
it 테스트 name 구체적으로 작성하여 수정
할 일을 삭제하면 해당 할 일이 삭제되는 테스트 추가

sejineeee committed:
review: useSelector 추가, BDD 적용

useSelector 추가
describe - it 구문 사용하여 BDD 적용하여 수정

sejineeee committed:
review: action type이 없을 경우 테스트 추가

action의 type이 없을 경우의 테스트 추가하였으나
action의 type을 읽을 수 없다고 테스트 실패

sejineeee commented:
## 내용
안녕하세요!!!
테스트용 데이터를 fixtures 파일로 분리하였습니다! (분명 지난주에 배웠는데 잊어버리다니...🥺 계속 생각하도록 노력하겠습니다!)
coverage가 100% 달성하지 못한 부분을 일단 수정하였습니다.
reducer 파일에서 사용했던 if문은 객체 리터럴을 사용하는 방법으로 수정하였습니다.
reducer 함수에서는 action type을 확인하는 부분을 삼항연산자로 적어보았습니다.
(삼항연산자보다 if문을 사용하는 게 좀 더 명시적이어서 좋을까요?)
일단 다른 부분은 coverage가 100% 달성하였으나, reducer 파일에 action type이 일치하는게 없을 경우 state를 반환하는 부분이 계속 실패합니다. 일단 멘토님에게 질문 남겨놓고, 그 후에도 다시 시도해보겠습니다!

## 질문
<img width="688" alt="스크린샷 2022-11-03 오후 8 58 06" src="https://user-images.githubusercontent.com/86041335/199716333-2ca420be-c162-4e2a-bf23-06cc3af8a7d4.png">

`reducer.test.jsx` 파일에서 action type이 없을 경우 test 부분에서 type을 읽을 수 없다고 하는데, 어떤 식으로 테스트를 접근해야 될 지 잘 모르겠습니다.

daadaadaah commented on src/ListContainer.test.jsx:
```diff
@@ -1,24 +1,36 @@
+  useSelector.mockImplementation((selector) => selector({ tasks }));
 
+  it('ListContainer를 렌더링한다.', () => {
     const { getByText } = render(<ListContainer />);
```
테스트 코드도 리팩터링하는 것이 좋습니다! 중복되는 코드들을 한번 정리해봅시다!ㅎㅎ

daadaadaah commented on src/reducer.js:
```diff
@@ -4,36 +4,34 @@ const initialState = {
+};
+
+export default function reducer(state = initialState, action) {
+  return reducers[action.type] ? reducers[action.type](state, action) : state;
```
sejineeee님이 작성해주 신 코드는`reducers[action.type]` 자체가 없을 때이므로, sejinneeee의 의도("reducer 파일에 action type이 일치하는게 없을 경우 state를 반환하는 부분이 계속 실패합니다.")에 맞게 표현된 코드가 아닙니다! if문을 활용해서 `reducers[action.type]` 자체가 없을 때 뿐만 아니라 action type이 없을 때도 함께 고려해서 작성해봅시다!ㅎㅎ

daadaadaah commented on src/reducer.test.js:
```diff
@@ -81,4 +81,14 @@ describe('reducer', () => {
     });
   });
+
+  it('action type이 일치하는 게 없을 경우 현재 상태를 반환한다', () => {
```
1. 이 테스트로 알고 싶은게, 정의된 액션 타입에 아닌 경우, 현재 상태를 반환하는 걸 테스트 해보고 싶으셨나요? 아니면 action 자체가 없는 걸 테스트 하고 싶으셨나요? 이것도 sejinneee님이 표현하신 테스트 내용과 테스트 코드가 일치하지 않습니다!
2. 만약, 정의된 액션 타입이 아닌 경우에 대해 테스트 해보고 싶으셨다면, reducer에 action 값을 넣어주셔야 됩니다!
```js
  describe('non-exited action type', () => {
    it('happens nothing', () => {
      const prevState = {
        newId: 100,
        taskTitle: '',
        tasks: [],
      };

      const action = {
        type: 'non-exited action type', // 값은 제가 임의로 넣었습니다!
      };

      const newState = reducer(prevState, action);

      expect(newState).toBe(prevState);
    });
  });
```

sejineeee replied:
action의 type이 일치하지 않은 경우를 표현하고 싶었습니다!! 처음에 리덕스가 동작될 때, type에 이러한 것이 들어오기 때문에 처리를 action의 type이 일치하지 않은 경우를 처리해주고 싶었습니다!
<img width="659" alt="스크린샷 2022-11-04 오후 4 20 19" src="https://user-images.githubusercontent.com/86041335/199923362-d7da0621-bfca-49a1-a5bc-c57685070a4d.png">

daadaadaah reviewed (changes requested):
질문 주신 내용 관련 코드에 코멘트 남겨 놨습니다! 참고하세요😄

sejineeee committed:
review: 중복된 코드 함수로 분리

중복된 render 함수를 분리하여 수정

sejineeee committed:
review: reducer 함수에 if문으로 수정, 조건 추가

삼항연산자를 사용한 조건문을 if문으로 수정
action이 없는 경우, action type이 일치하지 않는 경우로 조건 변경

sejineeee committed:
review: 기본 state를 반환하는 action test 수정

action의 type이 일치하지 않거나 없는 경우에 대한 test 추가
acdtion이 없는 경우에 대한 test 수정

sejineeee committed:
review: 아무런 동작을 하지 않는 경우 수정

아무런 동작을 하지 않을 때, reducer에 아무것도 전달하지 않도록 수정

sejineeee committed:
style: 중괄호 공백 추가

중괄호 공백 없음에서 있음으로 추가하여 수정

sejineeee commented:
## 내용
안녕하세요!! 멘토님의 코멘트 덕분에
-  reducer에서 deleteTask에 task에 존재하지 않는 경우일 때, id를 음수로 변경하였습니다.
- 중복된 코드를 함수로 분리하였습니다.
- action이 없을 경우를 조건문에 추가하였습니다.
- 테스트에 아무런 동작을 하지 않을 경우와 action의 type이 일치하지 않는 경우를 나뉘어 테스트하였습니다.

## 질문
1. reducer에서 조건을 체크해줬을 때, 옵셔널 체이닝을 이용해도 괜찮은가요? 아니면 if문을 사용하는 게 더 괜찮은가요?
```javascript
export default function reducer(state, action) {
  return reducers[action?.type] ? reducers[action.type](state, action) : state;
}
```
2. 코멘트를 참고하여 action의 type이 없는 경우를 체크해주고 싶어서, `action.hasOwnProperty('type')`으로 표현하려고 했는데, eslint에서 사용할 수 없다고 안내를 해주던데, 그 이유가 궁금해요...🤔

3. console을 출력해봤을 때, 아무런 행동을 해주지 않아도 기본적으로 action의 type이 들어와서, 일치하지 않기 때문에 undefined가 나와서 처음에 조건문을 action type만 확인하는 식으로 처리한 것이었는데, 이 부분이 아직 완전히 이해가 잘 되지 않습니다ㅠㅠ

이번 한 주도 감사합니다.

daadaadaah reviewed (approved):
> 1. reducer에서 조건을 체크해줬을 때, 옵셔널 체이닝을 이용해도 괜찮은가요? 아니면 if문을 사용하는 게 더 괜찮은가요?

- 개발자마다 다르겠지만, 저는 좀더 의도를 드러낼 수 있도록 if문을 사용하는게 읽기 편합니다!

> 2. 코멘트를 참고하여 action의 type이 없는 경우를 체크해주고 싶어서, action.hasOwnProperty('type')으로 표현하려고 했는데, eslint에서 사용할 수 없다고 안내를 해주던데, 그 이유가 궁금해요...🤔
- eslint 에서 어떤 메시지를 보여주면서 안된다고 했는지, 같이 보여주시면 좋을 것 같아요!

> 3. console을 출력해봤을 때, 아무런 행동을 해주지 않아도 기본적으로 action의 type이 아래 사진과 같이 들어와서 확인이 되길래, 처음에 조건문을 action type만 확인하는 식으로 처리한 것이었는데, 조건문에 action.type만 확인하면 안되는 것인가요...? 아직 이 부분이 완전하게 이해되지 않아서 궁금합니다!!

- 질문을 정리하자면, action.type만 확인하면 안된다는 의미가 아래 코드처럼 작성하면 안된다는 의미로 이해하면 될까용?
```
reducers[action.type] ? reducers[action.type](state, action) : state;
```

sejineeee commented:
제가 처음에 action에서 type을 확인하고자 할 때
![스크린샷 2022-11-05 오후 1 18 42](https://user-images.githubusercontent.com/86041335/200101201-3481aedf-5761-4d02-b059-8217c7c0daf2.png)
이런식으로 eslint 안내가 뜨길래 안내 문구를 찾아서, 
![스크린샷 2022-11-05 오후 1 30 17](https://user-images.githubusercontent.com/86041335/200101640-b12c1adf-ff10-4686-9a20-8499437f2264.png)
이렇게 바꿔서 해결해보았습니다!

세 번째 마지막 질문에서는 test에 대한 것이었어요!! 브라우저에서는 잘 작동했는데, 테스트에서는 계속 action.type을 찾을 수 없다고 했는데, 생각해보니 제가 테스트를 잘못 작성했던 것이었어요! (daadaadaah 멘토님께서 type을 체크해주고 싶을 때의 예시로 작성해주신 코드를 이용하면 되는 것이었더라고요...😯)

이번 한 주도 정말 감사합니다!! 두 번째 과제도 늦게나마 PR 올려보도록 하겠습니다!!!!
